In [2]:
import pandas as pd
import scipy.stats as stats

df = pd.read_csv(r'C:\Users\ferzi\python practice\Risk on a Loan Portfolio Using Statistical Tests\loan-dataset-1-clean.csv')
df

Unnamed: 0                               loan_id  \
0               0  70280c1b-4339-4949-ba69-ad2a6cb6b782   
1               1  8a550559-9749-491b-b6aa-7188bc0a488f   
2               2  5644d27e-e05e-4a40-bef3-64187892338f   
3               3  9d8d17ac-8af9-4936-ba5d-8d5ed5e0cfab   
4               4  4586ad7a-2231-419b-b8e3-ba03b929b56e   
...           ...                                   ...   
14995       15158  1a3c6a20-4da0-422c-8c89-d9d1906b2f84   
14996       15159  c870947f-28f3-424e-89d4-ebc0d8126bcb   
14997       15160  fde0b44b-f82b-4452-8719-8190e95a5884   
14998       15161  91bd0c9b-306c-41ea-b6b1-aa7b01c71b5c   
14999       15162  0608c5bc-2eaa-42c2-a034-3d189fda6567   

       principal_loan_amount request_date  interest_rate    loan_channel  \
0                       8880   2023-03-06            9.0             WEB   
1                      13720   2023-05-10           10.0      MOBILE_APP   
2                       2090   2023-02-16            7.5           AGENT   
3                       6820   2023-02-26            8.5      MOBILE_APP   
4                      13770   2023-04-13           10.0       AFFILIATE   
...                      ...          ...            ...             ...   
14995                   3320   2023-05-03            8.0           AGENT   
14996                  12370   2023-01-03            9.5  DIRECT_SELLING   
14997                  14150   2023-05-19           10.0       AFFILIATE   
14998                   3250   2023-01-22            8.0  DIRECT_SELLING   
14999                  10990   2023-01-21            9.5       AFFILIATE   

       is_approved  is_default  customer_monthly_income  customer_age  \
0             True       False                     6100            30   
1             True       False                     5150            39   
2             True       False                     6030            43   
3            False       False                     2150            54   
4            False       False                     5810            28   
...            ...         ...                      ...           ...   
14995         True       False                     3900            49   
14996         True       False                     3120            33   
14997         True       False                     3120            31   
14998         True       False                     6390            38   
14999         True       False                     3090            43   

      customer_gender  customer_credit_score  customer_zip_code  
0                   M                      3              10007  
1                   F                      3              10004  
2                   F                      3              10003  
3                   M                      0              10004  
4                   F                      2              10007  
...               ...                    ...                ...  
14995               M                      5              10005  
14996               M                      3              10009  
14997               F                      4              10007  
14998               M                      3              10009  
14999               M                      3              10005  

[15000 rows x 13 columns]

In [6]:
'''Evaluate The Impact of Interest Rates on Loan Default Rates

You have the following hypotheses to test:

Null hypothesis (from the operations team) : the interest rate has no impact on the loan default rate

Alternative hypothesis : interest rate does have an impact on the loan default rate.

Use the two-tailed t-test to evaluate the impact of interest rates on loan default rates. Split the data into two groups based on the interest rate levels, then compare the mean default rates between the two groups to analyze the hypothesis.
'''

median = df['interest_rate'].median()
     
high_interest = df[df['interest_rate'] > median]
low_interest = df[df['interest_rate'] <= median]

print(median)

high_default_rate = high_interest['is_default'].mean()
low_default_rate = low_interest['is_default'].mean()

print(high_default_rate)
print(low_default_rate)

9.0
0.03367684946632315
0.037424210746393474


In [7]:
t_stat, p_val = stats.ttest_ind(high_interest['is_default'], low_interest['is_default'])
print(t_stat)
print(p_val)

-1.1830953524322996
0.23679011271197217


In [9]:
print('High interest rate default rate: {:.2%}'.format(high_default_rate))
print('Low interest rate default rate: {:.2%}'.format(low_default_rate))
print('Two-tailed t-test: t-statistic = {:.2f}, p-value = {:.2f}'.format(t_stat, p_val))

High interest rate default rate: 3.37%
Low interest rate default rate: 3.74%
Two-tailed t-test: t-statistic = -1.18, p-value = 0.24


The t-statistic is -1.18 meaning that the average (mean) default rate for loans with high interest rates is slightly lower than the mean default rate for loans with low interest rates.

However, the p-value is 0.24, which is higher than the commonly used threshold of 0.05 for statistical significance.

This means that the difference in default rates between the two groups may not be statistically significant. In other words, the interest rate has no significant impact on loan default rate.


Hence the null hypothesis (the interest rate has no impact on the loan default rate) is true

In [11]:
import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm

#calculating the mean approval rating for the customer income

income_groups = pd.cut(df['customer_monthly_income'], bins=[0, 2500, 5000, 7500, np.inf])
approval_rate = df.groupby(income_groups)['is_approved'].mean()

0        (5000.0, 7500.0]
1        (5000.0, 7500.0]
2        (5000.0, 7500.0]
3           (0.0, 2500.0]
4        (5000.0, 7500.0]
               ...       
14995    (2500.0, 5000.0]
14996    (2500.0, 5000.0]
14997    (2500.0, 5000.0]
14998    (5000.0, 7500.0]
14999    (2500.0, 5000.0]
Name: customer_monthly_income, Length: 15000, dtype: category
Categories (4, interval[float64, right]): [(0.0, 2500.0] < (2500.0, 5000.0] < (5000.0, 7500.0] < (7500.0, inf]]
customer_monthly_income
(0.0, 2500.0]       0.781103
(2500.0, 5000.0]    0.776105
(5000.0, 7500.0]    0.775494
(7500.0, inf]            NaN
Name: is_approved, dtype: float64


In [13]:
'''Prepare the data for analysis:

    Use is_approved for dependent variable (variable to explain)

    Use customer_monthly_income for independent variable (variable to explain the variation in the dependent variable)

    Use Logit function to fit a logistic regression model to the data.
'''
X = sm.add_constant(df['customer_monthly_income'])
y = df['is_approved']
model = sm.Logit(y, X).fit()
print(model.summary())

Optimization terminated successfully.
         Current function value: 0.531339
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:            is_approved   No. Observations:                15000
Model:                          Logit   Df Residuals:                    14998
Method:                           MLE   Df Model:                            1
Date:                Tue, 22 Aug 2023   Pseudo R-squ.:               8.514e-06
Time:                        02:04:37   Log-Likelihood:                -7970.1
converged:                       True   LL-Null:                       -7970.2
Covariance Type:            nonrobust   LLR p-value:                    0.7126
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                       1.2688      0.067     18.948      0.000       1.138   

The coefficient for the customer_monthly_income is -5.539e-06, which means that one-unit increase in the customer_monthly_income results in a -5.539e-06 unit decrease in the log odds of loan approval (is_approved).

The p-value for this coefficient is 0.713, which is greater than the significance level of 0.05. 

Therefore, we fail to reject the null hypothesis that there is no linear relationship between income level and loan approval rates.

In [14]:
'''Find Variable(s) That Affect Loan Default Rate

Since interest rate does not impact the default rate, you need to find out possible variable(s) that affect loan default rate.

The risk team suggested the following variable possibilities:

    customer_monthly_income

    customer_zip_code

You need to use regression test against data to find out whether the risk team's suggestion is correct.

Prepare the data for analysis:

    Use is_default for the dependent variable (variable to explain)

    Use customer_monthly_income and customer_zip_code for independent variables (variable to explain the variation in the dependent variable)

    Use Logit function to fit a logistic regression model to the data.
'''

X = df[['customer_monthly_income', 'customer_zip_code']]
X = sm.add_constant(X)
y = df['is_default']
model = sm.Logit(y, X).fit()
print(model.summary())

         Current function value: 0.045447
         Iterations: 35
                           Logit Regression Results                           
Dep. Variable:             is_default   No. Observations:                15000
Model:                          Logit   Df Residuals:                    14997
Method:                           MLE   Df Model:                            2
Date:                Tue, 22 Aug 2023   Pseudo R-squ.:                  0.7072
Time:                        02:10:53   Log-Likelihood:                -681.70
converged:                      False   LL-Null:                       -2328.5
Covariance Type:            nonrobust   LLR p-value:                     0.000
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                    3.684e+04   1654.525     22.266      0.000    3.36e+04    4.01e+04
customer_monthly_income  -

C:\Users\ferzi\anaconda3\lib\site-packages\statsmodels\base\model.py:604: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


Examine the regression result

    p-value for customer_monthly_income is greater than the significance level of 0.05, which means customer_monthly_income does not affect default rate

    p-value for customer_zip_code is less than the significance level of 0.05, which means customer_zip_code significantly affect default rate

    However, warning message Maximum Likelihood optimization failed to converge suggests that the model may not have converged thus the results should be interpreted with caution

    The warning Possibly complete quasi-separation indicates that there may be perfect prediction or separation of the data points (in this case customer_zip_code perfectly affects default rate), so the results should be interpreted with caution

In [15]:
'''The operations team came up with these additional hypotheses:

Null hypothesis: loan channel has no impact on the loan approval rate

Alternative hypothesis: loan channel impacts the loan approval rate

Use ANOVA test to determine if there is a significant difference between loan default rates across different loan channels. If the test indicates that there is a significant difference, you can use post-hoc tests to determine which specific loan channels have different loan default rates.

Perform the ANOVA test. Specify is_default as the dependent variable, and loan_channel is the categorical independent variable.

Calculate ANOVA table for the regression model using the anova_lm function from statsmodels.stats.anova. Use Type II ANOVA to determine the significance of each independent variable in the presence of other independent variables.
'''
from statsmodels.formula.api import ols

# Convert the is_default to numeric to conform with ANOVA test requirement
df['is_default'] = df['is_default'].astype(int)

formula = 'is_default ~ C(loan_channel)'
model = ols(formula, data=df).fit()
aov_table = sm.stats.anova_lm(model, typ=2)
print(aov_table)

                     sum_sq       df           F         PR(>F)
C(loan_channel)   29.296879      4.0  223.138301  2.286606e-186
Residual         492.191054  14995.0         NaN            NaN


The p-value is less than significance level of 0.05, means there is a statistically significant difference between the groups of loan_channel

In [16]:
# Perform post-hoc tests to determine which loan channel has different impact on default rate. 
# Use Tukey's Honestly Significant Difference (HSD) test on the data. 
# Use is_default as the response variable and loan_channel column as the predictor variable.

mc = sm.stats.multicomp.MultiComparison(df['is_default'], df['loan_channel'])
mc_results = mc.tukeyhsd()
print(mc_results.summary())

        Multiple Comparison of Means - Tukey HSD, FWER=0.05        
    group1         group2     meandiff p-adj   lower  upper  reject
-------------------------------------------------------------------
     AFFILIATE          AGENT  -0.0034 0.9505 -0.0161 0.0093  False
     AFFILIATE DIRECT_SELLING   0.0004    1.0 -0.0125 0.0133  False
     AFFILIATE     MOBILE_APP  -0.0036 0.9398 -0.0165 0.0092  False
     AFFILIATE            WEB   0.1059   -0.0  0.0934 0.1185   True
         AGENT DIRECT_SELLING   0.0038 0.9321 -0.0091 0.0167  False
         AGENT     MOBILE_APP  -0.0002    1.0 -0.0131 0.0126  False
         AGENT            WEB   0.1093   -0.0  0.0968 0.1218   True
DIRECT_SELLING     MOBILE_APP   -0.004 0.9195  -0.017  0.009  False
DIRECT_SELLING            WEB   0.1055   -0.0  0.0928 0.1182   True
    MOBILE_APP            WEB   0.1095   -0.0  0.0969 0.1222   True
-------------------------------------------------------------------


Examine the post-hoc test. If the reject column is true for a pair of loan channels, then you can reject the null hypothesis and conclude that there is a significant difference between the default rates of the loan channels.

In the post-hoc table, the reject column is true for the pairs of loan channels (AFFILIATE, WEB), (AGENT, WEB), (DIRECT_SELLING, WEB), and (MOBILE_APP, WEB). This suggests that the default rate for loans obtained through the WEB channel is significantly different from the default rates for loans obtained through the other channels.

Hence we can reject the null hypothesis, since WEB channel contributes to loan default rate